docs/advanced/UsageWithTypescript.md

```diff
@@ -0,0 +1,280 @@
+
+TypeScript has the potential to bring the following benefits to a Redux application:
+
+1. Type safety for reducers, state and action creators
+2. Easy refactoring of typed code
+3. A superior developer experience in a team environment
+
+## A Practical Example
+
+We will be going through a simplistic chat application to demonstrate a possible approach to include static typing. This chat application will have two reducers. The _chat reducer_ will focus on storing the chat history and the _system reducer_ will focus on storing session information.
+
+The full source code is available on [codesandbox here](https://codesandbox.io/s/w02m7jm3q7). Note that by going through this example yourself you will experience some of the benefits of using TypeScript.
+
+## Type Checking State
+
+Adding types to each slice of state is a good place to start since it does not rely on other types. In this example we start by describing the chat reducer's slice of state:
+
+```ts
+// src/store/chat/types.ts
+
+export interface Message {
+  user: string
+  message: string
+  timestamp: number
+}
+
+export interface ChatState {
+  messages: Message[]
+}
+```
+
+And then do the same for the system reducer's slice of state:
+
+```ts
+// src/store/system/types.ts
+
+export interface SystemState {
+  loggedIn: boolean
+  session: string
+  userName: string
+}
+```
+
+Note that we are exporting these interfaces to reuse them later in reducers and action creators.
+
+## Type Checking Actions & Action Creators
+
+We will be using TypeScript's enums to declare our action constants. [Enums](https://www.typescriptlang.org/docs/handbook/enums.html) allow us to define a set of named constants. Note that we are making a tradeoff here when we declare our types in a sperate file. In exchange for seperating our types into a seperate file, we get to keep our other files more focussed on their purpose. While this tradeoff can improve the maintainability of the codebase, it is perfectly fine to organize your project however you see fit.
+
+Chat Action Constants & Shape:
+
+```ts
+// src/store/chat/types.ts
+
+export enum ChatActions {
+  SendMessage = 'SEND_MESSAGE'
+}
+
+interface SendMessageAction {
+  type: ChatActions.SendMessage
+  payload: Message
+}
+
+export type ChatActionTypes = SendMessageAction
+```
+
+Note that you can use TypeScript's Union Type to express multiple actions like so:
+
+```ts
+export type ChatActionTypes = SendMessageAction | DeleteMessageAction
+```
+
+With these types declared we can now also type check chat's action creators. In this case we are taking advantage of TypeScript's inference to avoid verbosity:
+
+```ts
+// src/store/chat/actions.ts
+
+import { Message, ChatActions } from './types'
+
+export function sendMessage(newMessage: Message) {
+  return {
+    type: ChatActions.SendMessage,
+    payload: newMessage
+  }
+}
+```
+
+Side note: to appreciate TypeScript's inference, here is what a more verbose version would look like:
```

Comment:
This doesn't feel _too_ much more verbose.  Could maybe just emphasize that the return type is what's being inferred, with a commented-out version of the function declaration line shown right above the actual line of code for comparison?

Reply:
I prefer to add the return type myself. It makes it more obvious what that action will actually be created. This in turn makes it easier once you refactor the payload of that action. With the less verbose example you might get the error in an unrelated file or none at all. With explicit annotation you get the error message where it originates.

Reply:
I generally think inference is more of a personal preference than anything else, it's not better or worse. Some people like verbose and others don't 😄. 

> or none at all

Should basically only occur when you are not consuming a property that was typed. For instance if you remove a payload from the return of an action creator (that is being infered) that was never used in the first place.